lib/identity_cache.rb

```diff
@@ -190,7 +190,7 @@ def fetch_multi(*keys)
     end
 
     def with_fetch_read_only_records(value = true)
-      old_value = fetch_read_only_records
+      old_value = Thread.current[:identity_cache_fetch_read_only_records]
```

Comment:
The leak was here, after the first invocation `Thread.current[:identity_cache_fetch_read_only_records]` was reset to `true` instead of `nil`.